### DISCUSSION TITLE
Proposal: !! operator

### DISCUSSION THREAD

**Discussion by aaulmer** [dotnet/csharplang#3103] [Language Ideas]
In the same fashion of the __++__ or __--__ operators, there should be a __!!__ operator for the cases when we want to flip a boolean without having to self-reassign it. This is especially nice in practice when using long camel-cased variable names. Because instead of looking for a ! in the midst of a long code line, we can much more easily identify !! at the end and recognize its importance.

```csharp
bool longCamelCasedVariableName = false;

// ... later in code ...
longCamelCasedVariableName = !longCamelCasedVariableName;
```
This instead would be written as:
```csharp
bool longCamelCasedVariableName = false;

// .. later in code ...
longCamelCasedVariableName!!;
```

**Comment by KalleOlaviNiemitalo** [comment]
If it's just a `bool`, then you can already do this:

```C#
longCamelCasedVariableName ^= true;
```

**Comment by themeeman** [comment]
This is very much nothing like ++ and --. That would imply the -- was two negations.

**Comment by szalapski** [comment]
I would never want `!!` to flip a boolean; this would be very confusing to Javascript developers where it means something entirely opposite.

**Comment by canton7** [comment]
Don't forget that a `!` suffix is used to suppress nullability warnings. `!!` looks like "*really* suppress nullability warnings".

**Comment by KalleOlaviNiemitalo** [comment]
Since #186, C# 7.2 also allows `static void Toggle(ref this bool flag) { flag = !flag; }`. That won't work on by-value properties though, unless <https://github.com/dotnet/csharplang/issues/1677#issuecomment-401374695> is implemented.

**Comment by MgSam** [comment]
I'll one up you. I want a `!!!!` prefix operator that tells me when something is true. That way I can really be clear when I want something.

``` C#
if (!!!!areThereCookies()) eatCookies();
```

**Comment by Thaina** [comment]
While I don't agree with syntax at all I think it is interesting reason for the feature

Maybe, what if we could write a shorthand like this?

```C#
bool longCamelCasedVariableName = false;

// .. later in code ...
longCamelCasedVariableName = !; // just symbol ! for shorthand the whole left side variable

someClass.someProperty.someValue.someBoolean = !; // also work with deep variable
```

**Comment by Flutterish** [comment]
If you shout at a boolean it will flip its..! ...value...
